router/batchrouter/worker.go

```diff
@@ -28,259 +23,241 @@ func newWorker(partition string, logger logger.Logger, brt *Handle) *worker {
+		status   *jobsdb.JobStatusT
+		sourceID string
+		destID   string
+	}
+
+	var jobsToBuffer []jobEntry
+
+	// Process jobs and check for drain conditions
+	for _, job := range destinationJobs.jobs {
+		sourceID := gjson.GetBytes(job.Parameters, "source_id").String()
+		destinationID := destWithSources.Destination.ID
+		jobIDConnectionDetailsMap[job.JobID] = jobsdb.ConnectionDetails{
+			SourceID:      sourceID,
+			DestinationID: destinationID,
+		}
+		sourceFound := false
+		for _, s := range destWithSources.Sources {
+			if s.ID == sourceID {
+				sourceFound = true
+				break
 			}
-			if drain, reason := brt.drainer.Drain(
-				job.CreatedAt,
-				destWithSources.Destination.ID,
-				gjson.GetBytes(job.Parameters, "source_job_run_id").String(),
-			); drain {
-				status := jobsdb.JobStatusT{
-					JobID:         job.JobID,
-					AttemptNum:    job.LastJobStatus.AttemptNum + 1,
-					JobState:      jobsdb.Aborted.State,
-					ExecTime:      time.Now(),
-					RetryTime:     time.Now(),
-					ErrorCode:     routerutils.DRAIN_ERROR_CODE,
-					ErrorResponse: routerutils.EnhanceJSON([]byte(`{}`), "reason", reason),
-					Parameters:    []byte(`{}`), // check
-					JobParameters: job.Parameters,
-					WorkspaceId:   job.WorkspaceId,
-				}
-				// Enhancing job parameter with the drain reason.
-				job.Parameters = routerutils.EnhanceJSON(job.Parameters, "stage", "batch_router")
-				job.Parameters = routerutils.EnhanceJSON(job.Parameters, "reason", reason)
-				drainList = append(drainList, &status)
-				drainJobList = append(drainJobList, job)
-				if _, ok := drainStatsbyDest[destinationID]; !ok {
-					drainStatsbyDest[destinationID] = &routerutils.DrainStats{
-						Count:     0,
-						Reasons:   []string{},
-						Workspace: job.WorkspaceId,
-					}
-				}
-				drainStatsbyDest[destinationID].Count = drainStatsbyDest[destinationID].Count + 1
-				if !slices.Contains(drainStatsbyDest[destinationID].Reasons, reason) {
-					drainStatsbyDest[destinationID].Reasons = append(drainStatsbyDest[destinationID].Reasons, reason)
-				}
-			} else {
-				if _, ok := jobsBySource[sourceID]; !ok {
-					jobsBySource[sourceID] = []*jobsdb.JobT{}
+		}
+		if !sourceFound {
+			status := jobsdb.JobStatusT{
+				JobID:         job.JobID,
+				AttemptNum:    job.LastJobStatus.AttemptNum + 1,
+				JobState:      jobsdb.Aborted.State,
+				ExecTime:      time.Now(),
+				RetryTime:     time.Now(),
+				ErrorCode:     routerutils.DRAIN_ERROR_CODE,
+				ErrorResponse: routerutils.EnhanceJSON([]byte(`{}`), "reason", "source_not_found"),
+				Parameters:    []byte(`{}`),
+				JobParameters: job.Parameters,
+				WorkspaceId:   job.WorkspaceId,
+			}
+			job.Parameters = routerutils.EnhanceJSON(job.Parameters, "stage", "batch_router")
+			job.Parameters = routerutils.EnhanceJSON(job.Parameters, "reason", "source_not_found")
+			drainList = append(drainList, &status)
+			drainJobList = append(drainJobList, job)
+			drainJobList = append(drainJobList, job)
```

Comment:
only need to add it once
```suggestion
			drainJobList = append(drainJobList, job)
```